rasa/engine/caching.py

```diff
@@ -0,0 +1,450 @@
+        """Adds the output to the cache (see parent class for full docstring)."""
+        if self._is_disabled():
+            return
+
+        cache_dir, output_type = None, None
+        if isinstance(output, Cacheable):
+            cache_dir, output_type = self._cache_output_to_disk(output, model_storage)
+
+        try:
+            self._add_cache_entry(
+                cache_dir, fingerprint_key, output_fingerprint, output_type
+            )
+        except OperationalError:
+            if cache_dir:
+                shutil.rmtree(cache_dir)
+
+            raise
+
+    def _add_cache_entry(
+        self,
+        cache_dir: Optional[Text],
+        fingerprint_key: Text,
+        output_fingerprint: Text,
+        output_type: Text,
+    ) -> None:
+        with self._sessionmaker.begin() as session:
+            cache_entry = self.CacheEntry(
+                fingerprint_key=fingerprint_key,
+                output_fingerprint_key=output_fingerprint,
+                last_used=datetime.utcnow(),
+                rasa_version=rasa.__version__,
+                result_location=cache_dir,
+                result_type=output_type,
+            )
+            session.merge(cache_entry)
+
+    def _is_disabled(self) -> bool:
+        return self._max_cache_size == 0.0
+
+    def _cache_output_to_disk(
+        self, output: Cacheable, model_storage: ModelStorage
+    ) -> Tuple[Optional[Text], Optional[Text]]:
+        # Use `TempDirectoryPath` instead of `tempfile.TemporaryDirectory` as this
+        # leads to errors on Windows when the context manager tries to delete an
+        # already deleted temporary directory (e.g. https://bugs.python.org/issue29982)
+        with rasa.model.TempDirectoryPath(tempfile.mkdtemp()) as temp_dir:
+            tmp_path = Path(temp_dir)
+            try:
+
+                output.to_cache(tmp_path, model_storage)
+
+                logger.debug(f"Caching output of type '{type(output)}' succeeded.")
+            except Exception as e:
+                logger.error(
+                    f"Caching output of type '{type(output)}' failed with the "
+                    f"following error:\n{e}"
+                )
+                return None, None
+
+            output_size = rasa.utils.common.directory_size_in_mb(tmp_path)
+            if output_size > self._max_cache_size:
+                logger.debug(
+                    f"Caching result of type '{type(output)}' was skipped "
+                    f"because it exceeds the maximum cache size of "
+                    f"{self._max_cache_size} MiB."
+                )
+                return None, None
+
+            while (
+                rasa.utils.common.directory_size_in_mb(
+                    self._cache_location,
+                    filenames_to_exclude=[self._cache_database_name],
+                )
+                + output_size
+                > self._max_cache_size
+            ):
+                self._drop_least_recently_used_item()
```

Comment:
Can it also be that the `self._max_cache_size` is too small or the `output_size` is too big and dropping least recently used item couldn't help freeing up the space needed to store in cache?

Reply:
👍🏻 That's covered by [this line](https://github.com/RasaHQ/rasa/blob/c141b28ccda13d24b7a9ae8db65f3880aa9d0890/rasa/engine/caching.py#L315)

Reply:
👍